File: PRACTICAS/CAMBIAR_RUBY/COMPROBACIONES JAVA/src/civitas_p1/Sorpresa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import java.util.ArrayList;

/**
 *
 * @author angel
 */
public class Sorpresa {
    
    private TipoSorpresa tipo;
    private MazoSorpresas mazo;
    private Tablero tablero;
    private int valor;
    private String texto;
    
    Sorpresa(Tablero tablero,  TipoSorpresa tipo){
        init();
        this.tablero = tablero;
        this.tipo = tipo;
	texto = null;
    }
    
    Sorpresa(Tablero tablero,  TipoSorpresa tipo, int valor, String texto){
        init();
        this.tablero = tablero;
        this.tipo = tipo; 
        this.valor = valor;
        this.texto = texto;
    }
    
    Sorpresa(TipoSorpresa tipo,  int valor, String texto ){
        init();
        this.tipo = tipo; 
        this.valor = valor;
        this.texto = texto;
    }
    
    Sorpresa(TipoSorpresa tipo, MazoSorpresas mazo){
        init();
        this.tipo = tipo; 
        this.mazo = mazo;
	texto = null;
    }
    
    private void init(){
        valor = -1;
        mazo = null;
        tablero = null;
    }
    
    public Boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
        if (actual >= 0 && actual < todos.size())
            return true;
        else
            return false;
    }
    
    private void informe(int actual, ArrayList<Jugador> todos){
        Diario.instance.ocurreEvento("Se esta aplicando una sorpresa al Jugador " + todos.get(actual).getNombre());               
    }
    
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        
        if(tipo==TipoSorpresa.IRCARCEL){
            aplicarAJugador_irCarcel(actual,todos);
        }
        else if(tipo==TipoSorpresa.IRCASILLA){
            aplicarAJugador_irACasilla(actual,todos);
        }
        else if(tipo==TipoSorpresa.PAGARCOBRAR){
            aplicarAJugador_pagarCobrar(actual,todos);
        }
        else if(tipo==TipoSorpresa.PORCASAHOTEL){
            aplicarAJugador_porCasaHotel(actual,todos);
        }
        else if(tipo==TipoSorpresa.PORJUGADOR){
            aplicarAJugador_porJugador(actual,todos);
        }
        else if(tipo==TipoSorpresa.SALIRCARCEL){
            aplicarAJugador_salirCarcel(actual,todos);
        }
    }
    
    private void aplicarAJugador_irCarcel(int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).encarcelar(tablero.getCarcel());        
        }
    }
    
    private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            int casillaActual = todos.get(actual).getNumCasillaActual();
            int tirada = tablero.calcularTirada(casillaActual, valor);
            int new_pos = tablero.nuevaPosicion(casillaActual, tirada);
            todos.get(actual).moverACasilla(new_pos);
            tablero.getCasilla(new_pos).recibeJugador(actual,todos);
        }
    }
    
    private void aplicarAJugador_pagarCobrar (int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).modificarSaldo(valor);
        }
    }

    private void aplicarAJugador_porCasaHotel (int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).modificarSaldo(valor*(todos.get(actual).cantidadCasasHoteles()));
        }
    }

    private void aplicarAJugador_porJugador (int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            todos.get(actual).modificarSaldo((-1)*valor*(todos.get(actual).cantidadCasasHoteles()));
            Sorpresa p1 = new Sorpresa(TipoSorpresa.PAGARCOBRAR, valor*-1, "PagarCobrar");
        
            for (int i = 0; i < todos.size(); i++)
                if (i != actual)
                    p1.aplicarAJugador_pagarCobrar(i,todos);
        
            Sorpresa p2 = new Sorpresa(TipoSorpresa.PAGARCOBRAR, valor*(todos.size() - 1), "PagarCobrar");
            p2.aplicarAJugador_pagarCobrar(actual,todos);
        }
    }
         
    private void aplicarAJugador_salirCarcel (int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            Boolean tienenCarcel = false;
            for (int i = 0; i < todos.size() && !tienenCarcel; i++){
                if (i!=actual)
                    if (todos.get(i).tieneSalvoconducto())
                        tienenCarcel = true;
            }
            
            if(!tienenCarcel){
                    todos.get(actual).obtenerSalvoconducto(this);
                    salirDelMazo();
            }
            
            
        }
    }
    
    void salirDelMazo(){
        if (tipo == TipoSorpresa.SALIRCARCEL){
            mazo.inhabilitarCartaEspecial(this);
        }
    }
    
    void usada (){
        if (tipo == TipoSorpresa.SALIRCARCEL)
            mazo.habilitarCartaEspecial(this);
    }
    
    
    @Override
    public String toString(){
        return texto;
    }
    
    
}   

File: PRACTICAS/practica1/Civitas_Ruby/lib/sorpresa.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  
  class Sorpresa
    
    def initialize(tablero = nil,tipo = nil,valor = 0,texto = nil,mazo = nil)
      
      @tablero = tablero
      @tipo = tipo
      @valor = valor
      @texto = texto
      @mazo = mazo
      
    end
    
    
    def self.sorpresa(tablero, tipo, valor, texto)
      new(tablero,tipo,valor,texto)     
    end
    
    def self.sorpresa2(tipo, valor, texto) 
      new(tipo,valor,texto) 
    end
    
    def self.sorpresa3(tipo, mazo)
      new(tipo,mazo)
    end
    
    def jugador_correcto(actual, todos)
      if (actual >= 0 && actual < todos.size)
        return true
      else
        return false
      end
    end
    
    private
    def informe(actual, todos)
      Diario.instance.ocurre_evento("Se esta aplicando una sorpresa al jugador #{todos[i].nombre}")
    end
    
    public
    def aplicar_a_jugador(actual, todos)
      
      if @tipo == TipoSorpresa::IRCARCEL
        aplicar_a_jugador_ir_carcel(actual, todos)
      
      elsif @tipo == TipoSorpresa::IRCASILLA
        aplicar_a_jugador_ir_a_casilla
        
      elsif @tipo == TipoSorpresa::PAGARCOBRAR
        aplicar_a_jugador_pagar_cobrar
        
      elsif @tipo == TipoSorpresa::PORCASAHOTEL
        aplicar_a_jugador_por_casa_hotel
        
      elsif @tipo == TipoSorpresa::PORJUGADOR
        aplicar_a_jugador_por_jugador
        
      elsif @tipo == TipoSorpresa::SALIRCARCEL
        aplicar_a_jugador_salir_carcel
     
      end
      
    end
    
    private
    def aplicar_a_jugador_ir_carcel(actual, todos)
      if jugador_correcto(actual,todos)
        informe(actual, todos)
        todos[actual].encarcelar(tablero.carcel)
      end
    end
    
    
    def aplicar_a_jugador_ir_a_casilla(actual, todos)
      if jugador_correcto(actual,todos)
        informe(actual, todos)
        casilla_actual = todos[actual].num_casilla_actual
        tirada = @tablero.calcular_tirada(casilla_actual, @valor)
        new_pos = tablero.nueva_posicion(casilla_actual, tirada)
        todos[actual].mover_a_casilla(new_pos)
        #todos[actual].recibeJugador()
      end
     
    end
    
    def aplicar_a_jugador_pagar_cobrar(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].modificar_saldo(@valor)
      end
    end
    
    def aplicar_a_jugador_por_casa_hotel(actual, todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].modificar_saldo(@valor*todos[actual].cantidad_casas_hoteles)
      end
      
    end
    
    def aplicar_a_jugador_por_jugador(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].modificar_saldo((-1)*@valor*todos[actual].cantidad_casas_hoteles)
        p1 = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, @valor*(-1), "pagar_cobrar")
        
        
        for i in (0..todos.size - 1)
          if i != actual
            p1.aplicar_a_jugar_pagar_cobrar(i,todos)
          end
          
        end
        
        p2 = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, @valor*(todos.size - 1), "pagar_cobrar")
        p2.aplicar_a_jugar_pagar_cobrar(actual,todos)
      end
    end
    
    def aplicar_a_jugador_salir_carcel(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        @tienen_carcel = false
        for i in (0..todos.size - 1)
          if i != actual && todos[i].tiene_salvoconducto
            @tienen_carcel = true
          end
        end
        
        if !@tienen_carcel
          
          todos[actual].obtener_salvonconducto(self)
          salirDelMazo()
        end
        
      end
    end
    
    public
    def salir_del_mazo
      if @tipo == TipoSorpresa::SALIRCARCEL
        @mazo.inhabilitarcartaespecial(self)
      end
      
    end
    
    def usada
      if @tipo == TipoSorpresa::SALIRCARCEL
        mazo.habilitar_carta_especial(self)
      end
    end
    
    def to_string
      return @texto
    end
    
    
   
end

end